### PR TITLE
[tradfri] Add support for Air Purifier

### PR DESCRIPTION
**Pull request by vivienbo**
# Description

This PR adds support for Tradfri Air Purifier (Strakvind) into the tradfri binding.
It supports:
* changing fan mode, controlling LED and locking physical button.
* reading current fan speed, PM2.5 ppm reading, filter uptime, time remaining before next filer check, air quality reading (aligned with Tradfri app) and a Switch to signal when filter needs to be reviewed.

Note: you will still need to run the filter review process through the Tradfri Mobile App.

# Testing

Tested on OpenHAB 4.0.0-M1 and on 3.4.4.
* [JAR file for OpenHAB 3.4.4](https://github.com/vivienbo/openhab-addons/releases/download/tradfri-air-purifier-1.0.1/org.openhab.binding.tradfri-3.4.4-SNAPSHOT.jar)
* [JAR file for OpenHAB 4.0.0-SNAPSHOT](https://github.com/vivienbo/openhab-addons/releases/download/tradfri-air-purifier-1.0.1/org.openhab.binding.tradfri-4.0.0-SNAPSHOT.jar)

# Changelog
General matter:
* Added documentation disambiguation Tradfri vs Dirigera
Tradfri Air Purifier features:
* Added Tradfri Air Purifier - fanMode and fanSpeed
* Workable Tradfri Air Purifier basic implementation
* Tradfri: modified fanMode type and definition
* Tradfri Air Purifier: Added disableLed
* Tradfri Air Purifier: Added lockPhysicalButton
* Tradfri Air Purifier: Added airQuality data
* Tradfri Air Purifier: Added filterCheck data
* Tradfri Air Purifier: Added translations
* Tradfri Air Purifier: Added filter_uptime
* Tradfri Air Purifier: code optimization
* Documentation for supported Air Purifier channels

Related issues:
* #14816

### CODE REVIEW

**Comment by vivienbo** [comment]
Made some updates to the README to fix some formatting and content-splitting issues (8f17e0c)

**Comment by vivienbo** [comment]
Proactive fixes in 7f4accf:
* Fixes Alphabetical order of properties
* Fixes a typo in some names

*Note: Additions of the ^M endings in tradfri.properties is coming from execution of mvn spotless:check*

File: bundles/org.openhab.binding.tradfri/README.md
Comment (jlaur) [inline bundles/org.openhab.binding.tradfri/README.md:99]: Can you align the table part above also?

File: bundles/org.openhab.binding.tradfri/src/main/resources/OH-INF/i18n/tradfri_de.properties
Comment (jlaur) [inline bundles/org.openhab.binding.tradfri/src/main/resources/OH-INF/i18n/tradfri_de.properties:1]: Translations must be provided through Crowdin, so please revert the changes in this file. See https://www.openhab.org/docs/developer/utils/i18n.html#managing-translations

File: bundles/org.openhab.binding.tradfri/src/main/resources/OH-INF/i18n/tradfri_fi.properties
Comment (jlaur) [inline bundles/org.openhab.binding.tradfri/src/main/resources/OH-INF/i18n/tradfri_fi.properties:1]: Translations must be provided through Crowdin, so please revert the changes in this file. See https://www.openhab.org/docs/developer/utils/i18n.html#managing-translations

File: bundles/org.openhab.binding.tradfri/src/main/resources/OH-INF/i18n/tradfri_fr.properties
Comment (jlaur) [inline bundles/org.openhab.binding.tradfri/src/main/resources/OH-INF/i18n/tradfri_fr.properties:1]: Translations must be provided through Crowdin, so please revert the changes in this file. See https://www.openhab.org/docs/developer/utils/i18n.html#managing-translations

File: bundles/org.openhab.binding.tradfri/src/main/resources/OH-INF/i18n/tradfri_it.properties
Comment (jlaur) [inline bundles/org.openhab.binding.tradfri/src/main/resources/OH-INF/i18n/tradfri_it.properties:1]: Translations must be provided through Crowdin, so please revert the changes in this file. See https://www.openhab.org/docs/developer/utils/i18n.html#managing-translations

File: bundles/org.openhab.binding.tradfri/src/main/resources/OH-INF/thing/thing-types.xml
Comment (jlaur) [inline bundles/org.openhab.binding.tradfri/src/main/resources/OH-INF/thing/thing-types.xml:184]:
```suggestion
		<label>Air Purifier Device</label>
```

or perhaps just "Air Purifier"?

File: bundles/org.openhab.binding.tradfri/src/main/resources/OH-INF/i18n/tradfri.properties
Comment (jlaur) [inline bundles/org.openhab.binding.tradfri/src/main/resources/OH-INF/i18n/tradfri.properties:3]: It seems something went wrong with the formatting or character set of this file. In GitHub it looks like all line endings have changed.

File: bundles/org.openhab.binding.tradfri/src/main/java/org/openhab/binding/tradfri/internal/TradfriBindingConstants.java
Comment (jlaur) [inline bundles/org.openhab.binding.tradfri/src/main/java/org/openhab/binding/tradfri/internal/TradfriBindingConstants.java:245]:
```suggestion
    public static final Set<Integer> AIR_PURIFIER_FANMODE = Set.of(FAN_MODE_OFF, FAN_MODE_AUTO,
            FAN_MODE_SPEED1, FAN_MODE_SPEED2, FAN_MODE_SPEED3, FAN_MODE_SPEED4, FAN_MODE_SPEED5);
```

Comment (jlaur) [inline bundles/org.openhab.binding.tradfri/src/main/java/org/openhab/binding/tradfri/internal/TradfriBindingConstants.java:245]:
```suggestion
    public static final Set<Integer> AIR_PURIFIER_FANMODE = Set.of(FAN_MODE_OFF, FAN_MODE_AUTO,
            FAN_MODE_SPEED1, FAN_MODE_SPEED2, FAN_MODE_SPEED3, FAN_MODE_SPEED4, FAN_MODE_SPEED5);
```

File: bundles/org.openhab.binding.tradfri/src/main/java/org/openhab/binding/tradfri/internal/TradfriBindingConstants.java
Comment (jlaur) [inline bundles/org.openhab.binding.tradfri/src/main/java/org/openhab/binding/tradfri/internal/TradfriBindingConstants.java:60]:
```suggestion
    public static final Set<ThingTypeUID> SUPPORTED_AIR_PURIFIER_TYPES_UIDS = Set.of(THING_TYPE_AIR_PURIFIER);
```

File: bundles/org.openhab.binding.tradfri/src/main/resources/OH-INF/thing/thing-types.xml
Comment (jlaur) [inline bundles/org.openhab.binding.tradfri/src/main/resources/OH-INF/thing/thing-types.xml:299]: Consider adding a category:
```suggestion
		<description>The duration for which the current filter was used.</description>
		<category>Time</category>
```

File: bundles/org.openhab.binding.tradfri/src/main/resources/OH-INF/thing/thing-types.xml
Comment (jlaur) [inline bundles/org.openhab.binding.tradfri/src/main/resources/OH-INF/thing/thing-types.xml:281]: Consider adding a category:
```suggestion
		<description>The number of minutes before the next filter check (or time since check is required if below 0).</description>
		<category>Time</category>
```

File: bundles/org.openhab.binding.tradfri/src/main/resources/OH-INF/thing/thing-types.xml
Comment (jlaur) [inline bundles/org.openhab.binding.tradfri/src/main/resources/OH-INF/thing/thing-types.xml:234]: Consider adding a category:
```suggestion
		<description>Displays the current fan speed (0..50).</description>
		<category>Fan</category>
```

**Review by jlaur — CHANGES_REQUESTED** [review]
Thanks for your contribution! The code looks good, only a few minor comments.

File: bundles/org.openhab.binding.tradfri/src/main/resources/OH-INF/i18n/tradfri.properties
Comment (vivienbo) [reply bundles/org.openhab.binding.tradfri/src/main/resources/OH-INF/i18n/tradfri.properties:3]: Strangely enough this happened when running `mvn i18n:generate-default-translations` as per documentation.
My assumption is this is because I am on a Windows environment, thought I guess the `mvn i18n:generate-default-translations` should ensure a unified line ending is used irrespective of the development environment.
Not sure if this is a big but I will revert and add translations by hand instead of using this target.

File: bundles/org.openhab.binding.tradfri/src/main/resources/OH-INF/thing/thing-types.xml
Comment (vivienbo) [reply bundles/org.openhab.binding.tradfri/src/main/resources/OH-INF/thing/thing-types.xml:184]: Opted for the shortest form "Air Purifier". Thanks.

File: bundles/org.openhab.binding.tradfri/src/main/resources/OH-INF/thing/thing-types.xml
Comment (vivienbo) [reply bundles/org.openhab.binding.tradfri/src/main/resources/OH-INF/thing/thing-types.xml:301]: Thanks for the suggestion, added categories for Time and Fan wherever adequate.

File: bundles/org.openhab.binding.tradfri/src/main/resources/OH-INF/i18n/tradfri.properties
Comment (vivienbo) [reply bundles/org.openhab.binding.tradfri/src/main/resources/OH-INF/i18n/tradfri.properties]: Resolved by reverting to original file and adding translations without line ending issues

**Review by vivienbo — COMMENTED** [review]
Thanks a lot @jlaur for your code review, really appreciated.
I fixed everything according to your recommendation, built and tested it to confirm it is still totally functional. New test packages for 3.4.4 and 4.0.0 will be updated in the description.

File: bundles/org.openhab.binding.tradfri/README.md
Comment (jlaur) [inline bundles/org.openhab.binding.tradfri/README.md:103]:
```suggestion
| air_quality_pm25   | Number:Dimensionless | Density of Particulate Matter of 2.5μm, measured in ppm                                  |
```

File: bundles/org.openhab.binding.tradfri/README.md
Comment (jlaur) [inline bundles/org.openhab.binding.tradfri/README.md:105]:
```suggestion
| filter_check_next  | Number:Time   | Time in minutes before the next filter check if > 0, if < 0 you are late checking the filter |
```

File: bundles/org.openhab.binding.tradfri/README.md
Comment (jlaur) [inline bundles/org.openhab.binding.tradfri/README.md:107]:
```suggestion
| filter_uptime      | Number:Time   | Time elapsed since the last filter change, in minutes                                    |
```

File: bundles/org.openhab.binding.tradfri/README.md
Comment (jlaur) [inline bundles/org.openhab.binding.tradfri/README.md:93]: Optional, just noticed it was also like this before:
```suggestion
| color_temperature  | Dimmer        | Color temperature from 0% = cold to 100% = warm                                          |
```

File: bundles/org.openhab.binding.tradfri/README.md
Comment (jlaur) [inline bundles/org.openhab.binding.tradfri/README.md:94]: Optional, just noticed it was also like this before:
```suggestion
| color              | Color         | Full color                                                                               |
```

File: bundles/org.openhab.binding.tradfri/README.md
Comment (jlaur) [inline bundles/org.openhab.binding.tradfri/README.md:95]: Optional, just noticed it was also like this before:
```suggestion
| battery_level      | Number        | Battery level (in %)                                                                     |
```

File: bundles/org.openhab.binding.tradfri/README.md
Comment (jlaur) [inline bundles/org.openhab.binding.tradfri/README.md:96]: Optional, just noticed it was also like this before:
```suggestion
| battery_low        | Switch        | Battery low warning (<=10% = ON, >10% = OFF)                                             |
```

File: bundles/org.openhab.binding.tradfri/README.md
Comment (jlaur) [inline bundles/org.openhab.binding.tradfri/README.md:97]: Optional, just noticed it was also like this before:
```suggestion
| power              | Switch        | Power switch                                                                             |
```

File: bundles/org.openhab.binding.tradfri/README.md
Comment (jlaur) [inline bundles/org.openhab.binding.tradfri/README.md:98]: Optional, just noticed it was also like this before:
```suggestion
| position           | Rollershutter | Position of the blinds from 0% = open to 100% = closed                                   |
```

**Review by jlaur — CHANGES_REQUESTED** [review]
Before merging, perhaps you could have a look at these small findings in the README?

File: bundles/org.openhab.binding.tradfri/README.md
Comment (vivienbo) [reply bundles/org.openhab.binding.tradfri/README.md:93]: On this one: the version of the README in the current head of the git displays like this:
image

* lowercase "c" for "color temperature"
  * perhaps it has been capital "C" some time in the past, however this change is not a colateral of this PR 
* capital "t" for "The brightness" while all other texts in the table start (and are proposed to start for the new lines) with a lowercase letter.

So my proposal - unless there are project convetions for that - would be to align everything to lowercase in that column:
* no change on "color temperature" (keep it as in the head)
* lowercase first letter of all added descriptions
* change the capital "T" in "The brightness" (first line of the table) to lowercase "t"

Comment (vivienbo) [reply bundles/org.openhab.binding.tradfri/README.md:93]: I'll opt in for uppercase as this seems to be how it was historically done in this add-on.

Comment (jlaur) [reply bundles/org.openhab.binding.tradfri/README.md:93]: Most important is to have consistency. But I would also opt for uppercasing first letter in the descriptions, since they are basically sentences.

**Comment by vivienbo** [comment]
Hi @jlaur, I reviewed the additions to the documentation and standardization of the tables to:
* Fix Item Types
* Fix abreviations (min -> minutes)
* Standardize `description` column of the `Channel Type ID` table (first letter in uppercase)
* Align the `Channel Type ID` table
* Improve the `Channels` > `Air Purifier` descriptive paragraph
* Align the non-ZLL capabilities table to similar format as ZLL One (uppercase first letter of each word; Align capabilities labels to `Channel Type ID` labels for consistency)

**Review by vivienbo — COMMENTED** [review]
Reviewed all changes and submitted new commit with further enhancements after rewriting the Air Purifier documentation.

File: bundles/org.openhab.binding.tradfri/README.md
Comment (jlaur) [inline bundles/org.openhab.binding.tradfri/README.md:82]: Here and below, "a " should probably be removed the they are now mentioned as "channels" (plural):
```suggestion
* `fan_mode` and `fan_speed` channels, which allows for control of the fan and reading of the current speed.
```

File: bundles/org.openhab.binding.tradfri/README.md
Comment (jlaur) [inline bundles/org.openhab.binding.tradfri/README.md:86]: Perhaps for consistency?
```suggestion
* a `filter_uptime` channel, which represents the current time since last filter change.
```

**Review by jlaur — CHANGES_REQUESTED** [review]
The documentation update looks good. 👍 Only some nitty gritty last details.

File: bundles/org.openhab.binding.tradfri/README.md
Comment (vivienbo) [reply bundles/org.openhab.binding.tradfri/README.md]: very valid remarks, fixed

**Review by jlaur — APPROVED** [review]
LGTM. 👍 @kaikreuzer, @cweitkamp - you are both code owners for this binding, would you like to review?

**Comment by jlaur** [comment]
> LGTM. 👍 @kaikreuzer, @cweitkamp - you are both code owners for this binding, would you like to review?

@kaikreuzer, @cweitkamp - gentle ping. 🙂

**Comment by jlaur** [comment]
@kaikreuzer, @cweitkamp - I really just need to know if you'd like to have a look before merging - or not? 🙂

**Comment by kaikreuzer** [comment]
Thanks for the regular pings @jlaur, I somehow missed them. I'd say you can feel free to merge if you do not hear anything back after a ping within 2 weeks. :-)
Many thanks for reviewing it!

**Comment by kaikreuzer** [comment]
(and of course many thanks to @vivienbo for the contribution!)

**Pull request merged**
